### PR TITLE
core: reserve devices for exclusive OP-TEE usage

### PR DESCRIPTION
**Pull request by ldts**
Use the device tree to reserve devices away from the normal world. The
devices should be **comma separated strings** identifying the dt nodes to
reserve.

For example (a single driver in imx6ull would be a single string):
 CFG_CORE_RESERVE_DRIVERS="\"/soc/aips-bus@2200000/crypto@2280000\""

This initial implementation has strtok in boot.c for simplicity (clearly it doesn't belong there). 

If there is interest in having this feature merged in the project, please could you suggest a place where to locate the strtok function and I'll repost?

### CODE REVIEW

File: core/arch/arm/kernel/boot.c [hunk 1]
```diff
@@ -1067,6 +1067,108 @@ static int mark_tzdram_as_reserved(struct dt_descriptor *dt)
...
+		item[i] = strtok(NULL, "/");
+		if (!item[i])
+			break;
+	}
+
+	if (i >= ARRAY_SIZE(item))
+		return -1;
+
+	if (fdt_path_offset(dt->blob, driver) >= 0)
+		return 0;
```
Comment (etienne-lms) [inline core/arch/arm/kernel/boot.c:1123]: If the node exists, don't you want to change it to `status="disabled"; secure-status='okay";`?
If the node does not exist, why adding it with a disabled status? (I assume these changes target the FDT that will be used by non-secure world/bootloader).

**Review by etienne-lms — COMMENTED** [review]
Couldn't the platform (i.e. **plat-xxx/main.c**) or the drivers (when they initialize) call straight such a helper function `fdt_force_node_secure_exclusive(fdt, ...)`, rather then an build `CFG_xxx`? For example here, if **plat-imx/main.c** could call this function when `AIPS3_BASE` is defined and `CFG_FDT=y`.

File: core/arch/arm/kernel/boot.c (hunk 1 above)
Comment (ldts) [reply core/arch/arm/kernel/boot.c:1123]: @etienne-lms right, so 
* if the node exists, yes we do not add the node and just change the status to disabled and the secure-status to okay. That is what the code does AFAICS?
* if the node does not exist, we do create the node anyways just so anyone inspecting sysfs is aware that the secure world is using the device.

**Comment by Emantor** [comment]
> For example (a single driver in imx6ull would be a single string):
> CFG_CORE_RESERVE_DRIVERS=""/soc/aips-bus@2200000/crypto@2280000""

A note on these paths: unfortunately they are not stable. Changes in the past included the removal of leading zeros among other things. Even this path does not work with the latest kernel, since `aips-bus` has been renamed to `bus`. I strongly recommend using compatibles for this or requiring an include of the `__symbols__` node for DT overlays and than using the phandles.

**Comment by ldts** [comment]
> > For example (a single driver in imx6ull would be a single string):
> > CFG_CORE_RESERVE_DRIVERS=""/soc/aips-bus@2200000/crypto@2280000""
> 
> A note on these paths: unfortunately they are not stable. Changes in the past included the removal of leading zeros among other things. Even this path does not work with the latest kernel, since `aips-bus` has been renamed to `bus`. I strongly recommend using compatibles for this or requiring an include of the `__symbols__` node for DT overlays and than using the phandles.

right, it is a pain that paths are not stable but since we still have the code below I wasnt sure where we stood.
https://github.com/OP-TEE/optee_os/blob/master/core/drivers/imx_wdog.c#L101

ok

**Comment by Emantor** [comment]
> right, it is a pain that paths are not stable but since we still have the code below I wasnt sure where we stood.
> https://github.com/OP-TEE/optee_os/blob/master/core/drivers/imx_wdog.c#L101

Yep, thats what https://github.com/OP-TEE/optee_os/pull/4116 is for :grin:

**Comment by ldts** [comment]
I'll close this PR - as it stands is clearly not going anywhere. thanks for the comments and the pointers @Emantor

**Pull request closed without merging**